test/common/upstream/load_balancer_simulation_test.cc

```diff
@@ -81,26 +113,88 @@ TEST(DISABLED_LeastRequestLoadBalancerWeightTest, Weight) {
   }
 
-  for (const auto& weight : weight_to_percent) {
-    std::cout << fmt::format("weight:{}, percent:{}\n", weight.first, weight.second);
+  std::sort(observed_pcts.begin(), observed_pcts.end());
+  std::sort(params.expected_selection_probs.begin(), params.expected_selection_probs.end());
+  ASSERT_EQ(observed_pcts.size(), params.expected_selection_probs.size());
+  for (uint64_t i = 0; i < observed_pcts.size(); i++) {
+    EXPECT_NEAR(params.expected_selection_probs[i], observed_pcts[i], params.tolerance_pct);
   }
 }
 
+// Simulate weighted LR load balancer and verify expected selection probabilities.
+TEST(LeastRequestLoadBalancerWeightTest, Weight) {
+  LRLBTestParams params;
+
+  // No active requests or weight differences. This should look like uniform random LB.
+  params.num_hosts = 3;
+  params.num_subset_hosts = 1;
+  params.active_request_count = 0;
+  params.expected_selection_probs = {33.333, 33.333, 33.333};
+  params.weight = 1;
+  params.tolerance_pct = 1.0;
+  leastRequestLBWeightTest(params);
+
+  // Single host (out of 3) with lots of in-flight requests. Given that P2C will choose 2 hosts and
+  // take the one with higher weight, the only circumstance that the host with many in-flight
+  // requests will be picked is if P2C selects it twice.
+  params.num_hosts = 3;
+  params.num_subset_hosts = 1;
+  params.active_request_count = 10;
+  params.expected_selection_probs = {44.45, 44.45, 11.1};
+  params.weight = 1;
+  params.tolerance_pct = 1.0;
+  leastRequestLBWeightTest(params);
+
+  // Same as above, but with 2 hosts. The busy host will only be chosen if P2C picks it for both
+  // selections.
+  params.num_hosts = 2;
+  params.num_subset_hosts = 1;
+  params.active_request_count = 10;
+  params.expected_selection_probs = {25, 75};
+  params.weight = 1;
+  params.tolerance_pct = 1.0;
+  leastRequestLBWeightTest(params);
+
+  // Heterogeneous weights with no active requests. This should behave identically to weighted
+  // round-robin.
+  params.num_hosts = 2;
+  params.num_subset_hosts = 1;
+  params.active_request_count = 0;
+  params.expected_selection_probs = {66.66, 33.33};
+  params.weight = 2;
+  params.tolerance_pct = 1.0;
+  leastRequestLBWeightTest(params);
+
+  // Same as above, but we'll scale the subset's weight with active requests. With a default
+  // active_request_bias of 1.0, the subset host with a single active request will be cut in half,
+  // making both hosts have an identical weight.
+  params.num_hosts = 2;
+  params.num_subset_hosts = 1;
+  params.active_request_count = 1;
+  params.expected_selection_probs = {50, 50};
+  params.weight = 2;
+  params.tolerance_pct = 1.0;
+  leastRequestLBWeightTest(params);
+
+  // Same as above, but with 3 hosts.
+  params.num_hosts = 3;
+  params.num_subset_hosts = 1;
+  params.active_request_count = 1;
+  params.expected_selection_probs = {33.3, 33.3, 33.3};
+  params.weight = 2;
+  params.tolerance_pct = 1.0;
```

Comment:
Seems this value is always 1.0. Consider removing it from being a parameter, and set it to a const in the function.